drivers/periph_common/init.c

```diff
@@ -36,8 +45,23 @@
 #include "periph/hwrng.h"
 #endif
 
+
 void periph_init(void)
 {
+    /* initialize stdio first to allow DEBUG() during later stages */
+#ifdef CPU_NATIVE
```

Comment:
This 'ifdef' hell should not be here. After all, `periph_common` is a platform independent module and adding all this platform dependencies here does destroy this whole concept!

The better way I see is to move these calls to the common platform modules, calling these functions from right before `periph_init` is called (or not in e.g. case of native). This way there is no need for any hard to read ifdefs anywhere, just a `#ifdef MODULE_STDIO` in for each supported architecture...

Reply:
One general remark: It is very rarely seen that someone jumps in at the very end of a process and after a lot of work has been done and asks for dropping core concepts of this process. For good reason. Also a lot of people [tend to get very upset](https://mailarchive.ietf.org/arch/msg/tls/CzjJB1g0uFypY8UDdr6P9SCQBqA) about such things.

That said, the idea of this PR is to assure that stdio is available when the periphs are initialized, so that using `DEBUG()` there is a save thing to do. This solves a real [issue](https://github.com/RIOT-OS/RIOT/issues/10645), as people starting to `ENABLE_DEBUG` in periph initialization would like to debug that periph code and don't want to think about why the debug output does not appear in the console or even crashes the board.

By having the stdio initialization here centralized, it is enforced that `DEBUG()` is safe to do in periph initialization. If it is up to the board/cpu, inconsistent behavior can easily sneak back it. So this idea about making RIOT *less* hardware dependent.

How about this instead:

``` C
#ifndef HAVE_ARCH_STDIO_INIT
static inline void arch_stdio_init(void)
{
    stdio_init();
}
#endif

void periph_init(void)
{
#if defined(MODULE_STDIO_UART) || defined(MODULE_STDIO_RTT)
    arch_stdio_init();
#endif
    ....
}
```

This would allow to provide a architecture specific `void arch_stdio_init()` if needed and handle the hardware specific stuff there.

Reply:
> One general remark: It is very rarely seen that someone jumps in at the very end of a process

I am aware I and I honestly don't feel good about my behavior. I simply havn't seen this PR before (as I am not spending too much time on tracking PRs lately), and I stumbled upon this today - but when I see something like this I would simply feel bad about staying quit. So please accept my apologies.

On the topic at hand: I still disagree with putting this initialization into `periph_common`. `periph_common` is about peripheral, and STDIO is clearly not part of this. So just using `periph_common/init.c` as init trigger for some (critical, yes)) system module breaks with a global concept and is ergo always a bad idea!

So what would be needed is some generic 'pre-init' function reliably called by all platforms before `periph_common/init`, that can be used to bootstrap this kind of system modules. I don't see a large issue with doing such a thing, right?

Process wise I would be ok with merging this PR first (as it is already nicely tested), if someone promises to address this issue right afterwards.

Reply:
Maybe it is best to split off the mips32 changes, as any PR trying to reduce the hardware specific differences will depend on that. Maybe @francois-berder could open an PR for that, as he authored those commits anyway?

And maybe it would be a good idea to add get a general picture how the boot up process in an *ideal* RIOT board look like. Maybe a sequence diagram like [this one](https://api.riot-os.org/riot-netdev-rx.svg) in the [doc of the Network Device Driver API](https://api.riot-os.org/group__drivers__netdev__api.html) could help. In that diagram the hardware specific part could be color coded, as those will likely still differ in various ways due to specific hardware dependent requirements.

Here is the sequence diagram I linked above (sadly with broken style):

![seq_diagram](https://api.riot-os.org/riot-netdev-rx.svg)

Such a diagram might help to a lot to at least have new added boards/architectures behave in a similar way and might be a good starting point to refactor and work out the differences of the existing boards.